Add native spirv + native DEPENDS fixes

th0ma7 opened the pull request:
## Description

1. Add native `llvm-spirv`
2. Allow `NATIVE_DEPENDS` functionaliy
3. Simplify `cmake|rust|meson|etc` build makefiles to always re-use either `spksrc.cross-cc.mk` or `spksrc.native-cc.mk` removing a lot of code duplication
4. Rework logging facility using `pipefail` to be at the following makelevels (also enabling toolchain logging):
  - `spksrc.tc.mk`
  - `spksrc.cross|native-cc.mk`
  - `spksrc-kernel-modules.mk`
5. Status logging is now printed only once per dependencies and only right before its actual configure call ensuring logs actually matches work on-going.
6. Fixed `AVAILABLE_TOOLCHAINS` and in turn `SUPPORTED_ARCHS` and `LATEST_ARCHS` common variables to remove `qoriq-rust` entry
7. Enable `time` stats logging by default to capture time needed for package build:
```
===>  20251029-004239 MAKELEVEL: 01, PARALLEL_MAKE: max, ARCH: aarch64-6.2.4, NAME: python312 [END]
653.87user 168.66system 4:35.78elapsed 298%CPU (0avgtext+0avgdata 859516maxresident)k
220272inputs+5083040outputs (205major+35728198minor)pagefaults 0swaps
```

Fixes # <!--Optionally, add links to existing issues or other PR's-->

## Checklist

- [ ] Build rule `all-supported` completed successfully
- [ ] New installation of package completed successfully
- [ ] Package upgrade completed successfully (Manually install the package again)
- [ ] Package [functionality was tested](https://github.com/SynoCommunity/spksrc/wiki/Package-Update-Policy#tests-checks)
- [ ] Any needed [documentation](https://github.com/SynoCommunity/spksrc/wiki/Create-documentation) is updated/created


### Type of change

<!--Please use any relevant tags.-->
- [ ] Bug fix
- [ ] New Package
- [ ] Package update
- [x] Includes small framework changes
- [ ] This change requires a documentation update (e.g. Wiki)

th0ma7 commented:
@hgy59 can you generate and publish an updated llvm package online?  New version will now provide llvm-spirv which is essential if I later want to include vulkan acceleration.

Also, there's a few changes to the framework - I haven't found any regression, it actually solves the dependency management from native which wasn't working, but it could also impact build elsewhere... testing on-going on my end, your help would be appreciated before pushing this one in.

hgy59 commented:
> @hgy59 can you generate and publish an updated llvm package online? New version will now provide llvm-spirv which is essential if I later want to include vulkan acceleration.

Got it, but can we update to a newer revision (like v14.0.6 .. v14.0.16)? This would prevent an overwrite of the already prebuilt v14.0.5.

hgy59 commented:
@th0ma7 another hint (seems to not find the already built SPIRV-Tools)

end of `===>  Configuring for llvm`

```
-- LLD version: 14.0.5
-- Found PkgConfig: /usr/bin/pkg-config (found version "1.8.1")
-- Using SPIR-V Headers from
          /spksrc/native/llvm-14.0-build/work-native/install/usr/local
-- Checking for one of the modules 'SPIRV-Tools'
-- SPIRV-Tools not found; project will be built without --spirv-tools-dis support.
-- Registering Bye as a pass plugin (static build: OFF)
-- Configuring done
-- Generating done
```

hgy59 commented:
@th0ma7 it is built and available as https://github.com/SynoCommunity/spksrc/releases/download/native%2Fllvm/native-llvm-14.0.5.txz

but AFAICS it is lacking clang-tblgen, ...

th0ma7 commented:
>  I guess the ninja build path is wrong (in the MSG output only):
> 
> ```
> -- Build files have been written to: /spksrc/native/llvm-14.0-build/work-native/llvm.build
> ===>  Compiling for llvm
> ===>  - Ninja compile
> ===>  - Ninja build path = /spksrc/native/llvm-14.0-build/work-native/llvm-project-14.0.5.src//spksrc/native/llvm-14.0-build/work-native/llvm.build
> ```

@hgy59 your guess is good... and I am able to reproduce locally:
```
spksrc@spksrc:~/vulkan-videodrv/spksrc/native/llvm-14.0-build$ grep 'Ninja build path' build.out 
===>  - Ninja build path = /home/spksrc/vulkan-videodrv/spksrc/native/llvm-14.0-build/work-native/SPIRV-Headers-vulkan-sdk-1.4.328.1//home/spksrc/vulkan-videodrv/spksrc/native/llvm-14.0-build/work-native/SPIRV-Headers-vulkan-sdk-1.4.328.1/build
===>  - Ninja build path = /home/spksrc/vulkan-videodrv/spksrc/native/llvm-14.0-build/work-native/SPIRV-Tools-2025.4//home/spksrc/vulkan-videodrv/spksrc/native/llvm-14.0-build/work-native/SPIRV-Tools-2025.4/build
===>  - Ninja build path = /home/spksrc/vulkan-videodrv/spksrc/native/llvm-14.0-build/work-native/llvm-project-14.0.5.src//home/spksrc/vulkan-videodrv/spksrc/native/llvm-14.0-build/work-native/llvm.build
```

As for the tools: clang-tblgen, llvm-config, llvm-link, llvm-spirv, llvm-tblgen, opt:
```
spksrc@spksrc:~/vulkan-videodrv/spksrc/native/llvm-14.0-build/work-native/install/usr/local/bin$ ll clang-tblgen llvm-config llvm-link llvm-spirv llvm-tblgen opt
ls: cannot access 'clang-tblgen': No such file or directory
-rwxr-xr-x 1 spksrc spksrc  264432 Oct 15 11:56 llvm-config
-rwxr-xr-x 1 spksrc spksrc   91240 Oct 15 12:03 llvm-link
-rwxr-xr-x 1 spksrc spksrc 6333456 Oct 15 12:04 llvm-spirv
-rwxr-xr-x 1 spksrc spksrc 4243848 Oct 15 11:53 llvm-tblgen
-rwxr-xr-x 1 spksrc spksrc  383128 Oct 15 12:04 opt
```
I thought I had nailed them all, including `clang-tblgen` using `-DCLANG_INSTALL_TABLEGEN=ON` but seems that last one is still missing.

> BTW the build of native packages still does not create build*.log files.

I know, it hurts, I'm so much used to have them everywhere else... Let me add that while this isn't fully backed yet.  And thnx for testing (there may be another round later :)

hgy59 commented:
@th0ma7 updated version with sha256:22e11a29783e6fcc09343d9d8de724492f547d45be761fe17715e6bc86bb9843 uploaded (same url).

th0ma7 commented:
@mreid-tt and @hgy59 i was able to do a major simplification of our main call-up scripts to always use native|cross-cc.mk which will greatly help maintenance in the future.

I still have cross-go to complete, otherwise the remaining would be ready for another round of testing so any remaining issue can be fixed.

If you have a moment to shake bugs up on this pr would be much appreciated.

hgy59 commented:
@th0ma7 didn't you plan to update native/llvm-14.0 with the updated prebuild that includes spriv?

th0ma7 commented:
@hgy59 yes, but as soon i push this in it will initiate long builds... Thus trying to reduce useless builds while i can.

hgy59 commented:
@th0ma7 
and this is my change in llvm-14.0-build/Makefile to create txz archive:
```
	cd work-native && tar -cJf ../native-$(PKG_NAME)-$(PKG_VERS).txz ./install/*
```

hgy59 commented:
@th0ma7 triggered by the log outputs of `/bin/bash: line 3: @echo: command not found` I analyzed the log output of the latest builds.

So I found that the synocli-videodriver package is created two times.
for x64-6.2.4 the first build started at:
```
2025-10-19T21:53:18.5171523Z make: Entering directory '/github/workspace/spk/synocli-videodriver'
```
and the second build about 15 minutes later:
```
2025-10-19T22:08:21.4260755Z make: Entering directory '/github/workspace/spk/synocli-videodriver'
```

both finally created the package:
first:
```
2025-10-19T22:08:21.3959542Z (cd /github/workspace/spk/synocli-videodriver/work-x64-6.2.4 && tar cpf /github/workspace/spk/synocli-videodriver/../../packages/synocli-videodriver_x64-6.2.4_1.2-3.spk --group=root --owner=root package.tgz INFO scripts conf PACKAGE_ICON.PNG PACKAGE_ICON_256.PNG)
```
second:
```
2025-10-19T22:23:25.6240245Z (cd /github/workspace/spk/synocli-videodriver/work-x64-source && tar cpf /github/workspace/spk/synocli-videodriver/../../packages/synocli-videodriver_x64-6.2.4_1.2-3.spk --group=root --owner=root package.tgz INFO scripts conf PACKAGE_ICON.PNG PACKAGE_ICON_256.PNG)
```


the two builds have different work folders
`/github/workspace/spk/synocli-videodriver/work-x64-6.2.4`
vs
`/github/workspace/spk/synocli-videodriver/work-x64-source`


how can WORK_DIR change from x64-6.2.4 to x64-source ?

th0ma7 commented:
Yeah... some of the tiny changes ends-up being really disruptive... that work_dir thing is a mess honestly.  Found a few other issues, will require a few more cycles to get properly sorted out.

th0ma7 commented:
@hgy59 two things left to do:
1. fix `make all-supported` - there's a quirk with `toolchain/syno-qoriq-6.2.4-rust` that uses `spksrc.native-*.mk` that requires playing with
2. Add a "dummy" spksrc.status.mk so only one message of each dependency is being submitted to `status-build.log`.  When it runs through the dependency tree it re-runs into some dependencies thus re-printing its status in the log file (while nothing actually hapens) such as:
```
===>  20251022-002933 MAKELEVEL: 11, PARALLEL_MAKE: max, ARCH: qoriq-6.2.4, NAME: zlib
===>  20251022-002934 MAKELEVEL: 09, PARALLEL_MAKE: max, ARCH: qoriq-6.2.4, NAME: libpng
===>  20251022-002935 MAKELEVEL: 11, PARALLEL_MAKE: max, ARCH: qoriq-6.2.4, NAME: bzip2
===>  20251022-002936 MAKELEVEL: 11, PARALLEL_MAKE: max, ARCH: qoriq-6.2.4, NAME: libpng
===>  20251022-002936 MAKELEVEL: 11, PARALLEL_MAKE: max, ARCH: qoriq-6.2.4, NAME: zlib
===>  20251022-002936 MAKELEVEL: 09, PARALLEL_MAKE: max, ARCH: qoriq-6.2.4, NAME: freetype
===>  20251022-002939 MAKELEVEL: 11, PARALLEL_MAKE: max, ARCH: qoriq-6.2.4, NAME: libpng
```

Other than that its getting darn close to be ready.  And quite honesly, it fixes many quirks that got added over time while simplifying overall code significantly.

th0ma7 commented:
Thnx, much appreciated, I'll look into it.

th0ma7 commented:
@hgy59 Everything looks like fully functional and quite happy with things.

Now yet to be found: how to enable main/overall log file for kernel builds... still looking :shrug:

th0ma7 commented:
@hgy59 first off, much thnx for your multiple review iterations on this PR!

I have now fixed the logging for kernel builds... it allowed me to find that the toolchain part was also affected, now fixed.  output differs slightly but allow catching all the outputs.

I tested things quite extensively but as usual these sort of deeper framework changes always requires a few fixes afterwards.

Anyhow, if you have a few more cycles to spare to test things up would be greatly appreciated, further considering your Eye-Of-The-Tiger allowed identifing quite a few things up until now.

Otherwise, I consider this ready for merge (although pursuing testing in parallel in case there is anything else to be found)

hgy59 commented:
@th0ma7 currently building synokernel-cdrom for x64-7.1.

The build log starts with:
```
env make --no-print-directory cmake_vars > /spksrc/toolchain/syno-x64-7.1/work/tc_vars.cmake 2>/dev/null;
env make --no-print-directory meson_cross_vars > /spksrc/toolchain/syno-x64-7.1/work/tc_vars.meson-cross 2>/dev/null;
env make --no-print-directory meson_native_vars > /spksrc/toolchain/syno-x64-7.1/work/tc_vars.meson-native 2>/dev/null;
env make --no-print-directory tc_vars > /spksrc/toolchain/syno-x64-7.1/work/tc_vars.mk 2>/dev/null;
```

and contains those 4 lines for every specific arch it builds
```
===>  kernel-modules archs to be processed: apollolake-7.1 avoton-7.1 braswell-7.1 broadwell-7.1 broadwellnk-7.1 broadwellnkv2-7.1 broadwellntbap-7.1 bromolow-7.1 cedarview-7.1 denverton-7.1 geminilake-7.1 grantley-7.1 kvmx64-7.1 purley-7.1 r1000-7.1 v1000-7.1
```

and it creates a log file for each arch like
```
build-apollolake-7.1.log
build-avoton-7.1.log
...
```
And each of the file has exactly those 4 lines.

_`env make --no-print-directory` appears 68 times in build-x64-7.1.log, 4 times in each of the 16 build-specific_arch-7.1.log and 4 times in each of the 16 build-x84-7.1-kernel-modules-specific_arch-.7.1.log_


May be some behaviour is by intention for the kernel modules, built for each specific arch of the generic x64.
- a folder `tc_vars-backup` exists in the work folder so I guess the intention is that the `tc_vars.*` files are created only once for the generic arch and reused for the _specific_archs_ (but the log files tell, that this is done for each specific arch.
- I guess those 4 lines should not be printed in the build logs
- the build log files for specific kernel-module archs (with only the 4 lines) should not be created.

th0ma7 commented:
> IMHO we don't need logs of toolchain builds. In local environments the toolchains are built only once (and have to be rebuilt when build process or digests changed). In github builds toolchains are usually loaded from the cache and not built either.
> 
> Or you find another way to omit such 4 lines when toolchains are already built....

@hgy59 this behavior has not change from previously.  You see, to add logs to native (as you asked) I needed to move the logs away from `spksrc.supported.mk` and instead having them one level below, resulting in modifications to `spksrc.cross|native-cc.mk` (and ensuring every builds ends-up using them i.e. `spksrc.cross-cmake|meson|rust|etc.mk`).  In turn, besides to removing a lot of duplicated code, this disabled logs from the lower level i.e. toolchain - so to re-enable it I had to introduce logging within `spksrc.tc.mk` to reproduce the original behavior.

Why toolchain logs you may ask: well this allowed me to find long-standing issues with `SUPPORTED_ARCHS` and `LATEST_ARCHS` variables for instance which was processing `qoriq-rust` wrongly.  Could they be disabled: yes.  I could add a variable in local.mk such as `LOG_TOOLCHAIN|CROSS|NATIVE = no|false|disabled` or similar and test for that.  Although I don't see much benefit from it.

Side note: makes me think I should check the other toolchain variables from arch.mk to ensure they are also ok.

Unless you insist I would rather keep things as-is?

hgy59 commented:
> Unless you insist I would rather keep things as-is?

The problem is, that it was difficult to find the source of those 4 log lines.
I am used to see a context in the output with lines prefixed with `===>  `  in the log files, this is lacking here.

We should definitfly not add another option to disable logs...
keep it as simple as possible _(it's already far too complicated)_ ...

th0ma7 commented:
So if I read, no objection a merging this as-is?

th0ma7 commented:
@hgy59 yes, see comment above that explains sub-arch logging.
